### PR TITLE
PP-5402-Serialization/Deserialization of Telephone Payments

### PR DESCRIPTION
**Pull request by nimalank7**
## WHAT YOU DID
This PR creates classes for serializing and deserialization the JSON of Telephone Payments as well as representation tests which check whether they can successfully serialize/deserialize.

I created the TelephoneChargeRequest which deserializes the following JSON:

`{
  "amount": 12000,
  "reference": "MRPC12345",
  "description": "New passport application",
  "created_date": "2018-02-21T16:04:25Z",
  "authorised_date": "2018-02-21T16:05:33Z",
  "processor_id": "183f2j8923j8",
  "provider_id": "17498-8412u9-1273891239",
  "auth_code": "666",
  "payment_outcome": {
    "status": "success",
    "code": "P0010",
    "supplemental": {
      "error_code": "ECKOH01234",
      "error_message": "textual message describing error code"
    }
  },
  "card_type": "master-card",
  "name_on_card": "Jane Doe",
  "email_address": "jane_doe@example.com",
  "card_expiry": "02/19",
  "last_four_digits": "1234",
  "first_six_digits": "654321",
  "telephone_number": "+447700900796"
}`

I created the TelephoneChargeResponse which serializes to the following JSON:

`{
  "amount": 12000,
  "reference": "MRPC12345",
  "description": "New passport application",
  "created_date": "2018-02-21T16:04:25Z",
  "authorised_date": "2018-02-21T16:05:33Z",
  "processor_id": "183f2j8923j8",
  "provider_id": "17498-8412u9-1273891239",
  "auth_code": "666",
  "payment_outcome": {
    "status": "success",
    "code": "P0010",
    "supplemental": {
      "error_code": "ECKOH01234",
      "error_message": "textual message describing error code"
    }
  },
  "card_type": "master-card",
  "name_on_card": "Jane Doe",
  "email_address": "jane_doe@example.com",
  "card_expiry": "02/19",
  "last_four_digits": "1234",
  "first_six_digits": "654321",
  "telephone_number": "+447700900796",
  "payment_id": "hu20sqlact5260q2nanm0q8u93",
  "state": {
    "status": "success",
    "finished": true,
    "message": "Hello, world!",
    "code": "P0010"
  }
}`

## How to test

Please review the tests in the TelephonePaymentJSONTest.

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/charge/model/telephone/TelephoneChargeCreateRequest.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    
+    private String firstSixDigits;
+    
+    private String telephoneNumber;
+
+    public TelephoneChargeCreateRequest() {
+        // To enable Jackson serialisation we need a default constructor
+    }
+
+    public TelephoneChargeCreateRequest(int amount, String reference, String description, String createdDate, String authorisedDate, String processorId, String providerId, String authCode, PaymentOutcome paymentOutcome, String cardType, String nameOnCard, String emailAddress, String cardExpiry, String lastFourDigits, String firstSixDigits, String telephoneNumber) {
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/charge/model/telephone/TelephoneChargeCreateRequest.java]: Would be good to use builder here (also highlighted by Codacy)

File: src/main/java/uk/gov/pay/connector/charge/model/telephone/TelephoneChargeResponse.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                                    String cardType,
+                                    String nameOnCard,
+                                    String emailAddress,
+                                    String cardExpiry,
+                                    String lastFourDigits,
+                                    String firstSixDigits,
+                                    String telephoneNumber,
+                                    String paymentId,
+                                    State state) {
+        // For testing serialization
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/charge/model/telephone/TelephoneChargeResponse.java]: Would be good to use builder here (also highlighted by Codacy)

**Review by kbottla — COMMENTED** [review]
Looks good. Would be nice if Builder is used instead of constructor as highlighted by Codacy

**Review by kbottla — DISMISSED** [review]
Looks good. Couple of comments about using Builder but happy them to be addressed separately

File: src/main/java/uk/gov/pay/connector/charge/model/telephone/TelephoneChargeCreateRequest.java (hunk 1 above)
Comment (nimalank7) [reply src/main/java/uk/gov/pay/connector/charge/model/telephone/TelephoneChargeCreateRequest.java]: Done

File: src/main/java/uk/gov/pay/connector/charge/model/telephone/TelephoneChargeResponse.java (hunk 2 above)
Comment (nimalank7) [reply src/main/java/uk/gov/pay/connector/charge/model/telephone/TelephoneChargeResponse.java]: Done

File: src/main/java/uk/gov/pay/connector/charge/model/telephone/TelephoneChargeResponse.java [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+        this.cardExpiry = chargeBuilder.cardExpiry;
+        this.lastFourDigits = chargeBuilder.lastFourDigits;
+        this.firstSixDigits = chargeBuilder.firstSixDigits;
+        this.telephoneNumber = chargeBuilder.telephoneNumber;
+        this.paymentId = chargeBuilder.paymentId;
+        this.state = chargeBuilder.state;
+    }
+    
+    /*
+    public TelephoneChargeResponse(int amount,
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/charge/model/telephone/TelephoneChargeResponse.java]: Can this be removed ?

Comment (nimalank7) [reply src/main/java/uk/gov/pay/connector/charge/model/telephone/TelephoneChargeResponse.java]: Apologies - yep I've removed this!

**Review by kbottla — APPROVED** [review]

**Pull request merged**